pkg/kudoctl/util/repo/package.go

```diff
@@ -0,0 +1,312 @@
+	frameworkV0FileName = "-framework.yaml"
+	versionV0FileName   = "-frameworkversion.yaml"
+	instanceV0FileName  = "-instance.yaml"
+
+	frameworkV1FileName     = "framework.yaml"
+	templatesV1Folder       = "templates"
+	templateV1FileNameRegex = "templates/.*.yaml"
+	paramsV1FileName        = "params.yaml"
+)
+
+const apiVersion = "kudo.k8s.io/v1alpha1"
+
+// InstallCRDs is collection of CRDs that are used when installing framework
+// during installation, package format is converted to this structure
+type InstallCRDs struct {
+	Framework        *v1alpha1.Framework
+	FrameworkVersion *v1alpha1.FrameworkVersion
+	Instance         *v1alpha1.Instance
+}
+
+func isFrameworkV0File(name string) bool {
+	return strings.HasSuffix(name, frameworkV0FileName)
+}
+
+func isVersionV0File(name string) bool {
+	return strings.HasSuffix(name, versionV0FileName)
+}
+
+func isInstanceV0File(name string) bool {
+	return strings.HasSuffix(name, instanceV0FileName)
+}
+
+// ReadTarballPackage reads package from tarball and converts it to the CRD format
+func ReadTarballPackage(r io.Reader) (*InstallCRDs, error) {
+	p, err := untarV1Package(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "while untarring package")
+	}
+	return p.getInstallCRDs()
+}
+
+// ReadFileSystemPackage reads package from filesystem and converts it to the CRD format
+func ReadFileSystemPackage(path string) (*InstallCRDs, error) {
+	p, err := fromFilesystem(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "while reading package from filesystem")
+	}
+	return p.getInstallCRDs()
+}
+
+func fromFilesystem(packagePath string) (*v1Package, error) {
+	result := newV1Package()
+	err := filepath.Walk(packagePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			// skip directories
+			return nil
+		}
+		relativePath := strings.TrimPrefix(path, packagePath)
+		if path == packagePath {
+			// skip the root folder, as Walk always starts there
+			return nil
+		}
+		bytes, err := ioutil.ReadFile(path)
+		if err != nil {
+			return err
+		}
+		switch {
+		case isFrameworkV1File(info.Name()):
+			var bf bundle.Framework
+
+			if err = yaml.Unmarshal(bytes, &bf); err != nil {
+				return errors.Wrap(err, "cannot unmarshal framework")
+			}
+			result.Framework = &bf
+		case info.Name() == templatesV1Folder && info.IsDir():
+			// skip the folder itself, wait until we recursively start going into the template files
+			return nil
+		case isTemplateV1File(relativePath):
+			name := strings.TrimPrefix(relativePath, "/templates/")
+			result.Templates[name] = string(bytes)
+		case isParametersV1File(info.Name()):
+			if err = yaml.Unmarshal(bytes, &result.Params); err != nil {
+				return errors.Wrapf(err, "unmarshalling %s content", info.Name())
+			}
+		default:
+			return fmt.Errorf("unexpected file in the fileststem structure %s", info.Name())
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &result, nil
```

Comment:
questions that come to mind:
1. is there any file size limits? or any validation other than the rules of marshaling? 
(the concern is byte arrays and sizes)
2. there is no validation after the fact.   There seems to be a happy path of all the files should be there.   What if there is a framework file, but not other file?   or template file and no framework file?

Reply:
We validate when we do `getInstallCRDs`. You cannot really do it much earlier because some parts of the validation can be done only AFTER you load everything and you start parsing the loaded YAML files. E.g. to find out that you're missing a template file that is included inside the package. All that will be caught in `getInstallCRDs` and it will return an error

Reply:
On the size, we should at least validate the bundled package is not larger than 1MB (etcd's max key size). I know it's unlikely, but I'm not sure if we would get a sane/user understandable error back from Kubernetes.